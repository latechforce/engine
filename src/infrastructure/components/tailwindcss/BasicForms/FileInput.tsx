import { Context } from './Base'
import type { FileInputProps } from '/domain/components/BasicForms/FileInput'

// https://preline.co/docs/file-input.html

export const FileInput = ({
  field,
  label,
  description,
  placeholder,
  required,
  multiple,
}: FileInputProps) => {
  return (
    <Context label={label} description={description} field={field} required={required}>
      <label className="block">
        <span className="sr-only">{placeholder}</span>
        <input
          type="file"
          name={field}
          multiple={multiple}
          required={!!required}
          className="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400"
        />
      </label>
    </Context>
  )
}
