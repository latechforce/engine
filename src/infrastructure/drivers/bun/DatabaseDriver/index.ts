import type { IDatabaseDriver } from '/adapter/spi/drivers/DatabaseSpi'
import type { DatabaseConfig, DatabaseEventType } from '/domain/services/Database'
import type { EventDto } from '/adapter/spi/dtos/EventDto'
import { PostgreSQLDatabaseDriver } from '/infrastructure/drivers/common/DatabaseDriver/PostgreSQLDriver'
import { SQLiteDatabaseDriver } from './SQLite/SQLiteDriver'
import type { AutomationHistoryRecord } from '/domain/entities/Automation/History'
import type { TableConfig } from '/domain/entities/Table'
import type { TableSchema } from '../../../../adapter/api/schemas/TableSchema'
import { TableMapper } from '/adapter/api/mappers/TableMapper'

export class DatabaseDriver implements IDatabaseDriver {
  private _db: PostgreSQLDatabaseDriver | SQLiteDatabaseDriver
  public type: DatabaseConfig['type']

  constructor(config: DatabaseConfig) {
    this.type = config.type
    switch (this.type) {
      case 'SQLite':
        this._db = new SQLiteDatabaseDriver(config)
        break
      case 'PostgreSQL':
        this._db = new PostgreSQLDatabaseDriver(config)
        break
      default:
        throw new Error('Invalid driver')
    }
  }

  connect = async (): Promise<void> => {
    await this._db.connect()
  }

  disconnect = async (): Promise<void> => {
    await this._db.disconnect()
  }

  exec = async (query: string): Promise<void> => {
    await this._db.exec(query)
  }

  query = async <T>(
    text: string,
    values: (string | number | Buffer | Date)[] = []
  ): Promise<{ rows: T[]; rowCount: number }> => {
    return this._db.query(text, values)
  }

  table(table: TableConfig) {
    return this._db.table(table)
  }

  tableFromSchema(schema: TableSchema) {
    return this._db.table(TableMapper.toConfig(schema))
  }

  on = (event: DatabaseEventType, callback: (eventDto: EventDto) => void) => {
    this._db.on(event, callback)
  }

  setupTriggers = async (tables: string[]) => {
    await this._db.setupTriggers(tables)
  }

  waitForAutomationsHistories = async ({
    count = 1,
    status = 'succeed',
    timeout = 60000,
  }: {
    count?: number
    status?: string
    timeout?: number
  } = {}) => {
    let histories: AutomationHistoryRecord[] = []
    let retry = 0
    do {
      await new Promise((resolve) => setTimeout(resolve, 500))
      const { rows } = await this.query<AutomationHistoryRecord>(
        'SELECT * FROM automations_histories_view WHERE status = ?',
        [status]
      )
      histories = rows
      retry++
    } while (histories.length < count && retry < Math.floor(timeout / 500))
    if (histories.length < count) {
      throw new Error(`Automations histories not found: ${count} ${status} histories`)
    }
    return histories
  }
}
