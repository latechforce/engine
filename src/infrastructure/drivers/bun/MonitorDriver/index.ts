import type { MonitorConfig } from '/domain/services/Monitor'
import type { IMonitorDriver } from '/adapter/spi/drivers/MonitorSpi'
import { SentryDriver } from './SentryDriver'
import { ConsoleDriver } from '../../common/MonitorDriver/ConsoleDriver'

export class MonitorDriver implements IMonitorDriver {
  private _monitors: (SentryDriver | ConsoleDriver)[] = []

  constructor(config: MonitorConfig[]) {
    for (const monitor of config) {
      const { driver } = monitor
      switch (driver) {
        case 'Sentry':
          this._monitors.push(new SentryDriver())
          break
        case 'Console':
          this._monitors.push(new ConsoleDriver())
          break
        default:
          throw new Error('Invalid driver')
      }
    }
  }

  captureException = (error: Error) => {
    this._monitors.map((m) => m.captureException(error))
  }

  captureMessage = (message: string) => {
    this._monitors.map((m) => m.captureMessage(message))
  }
}
