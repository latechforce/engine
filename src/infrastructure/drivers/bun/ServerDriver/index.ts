import type { IServerDriver } from '/adapter/spi/drivers/ServerSpi'
import type { DeleteDto, GetDto, PatchDto, PostDto, RequestDto } from '/adapter/spi/dtos/RequestDto'
import type { ServerConfig, ServerMethodOptions } from '/domain/services/Server'
import type { Response } from '/domain/entities/Response'
import { ElysiaDriver } from './ElysiaDriver'

export class ServerDriver implements IServerDriver {
  private _server: ElysiaDriver

  constructor(config: ServerConfig) {
    this._server = new ElysiaDriver(config)
  }

  get = async (
    path: string,
    handler: (getDto: GetDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.get(path, handler, options)
  }

  post = async (
    path: string,
    handler: (postDto: PostDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.post(path, handler, options)
  }

  patch = async (
    path: string,
    handler: (patchDto: PatchDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.patch(path, handler, options)
  }

  delete = async (
    path: string,
    handler: (deleteDto: DeleteDto) => Promise<Response>,
    options: ServerMethodOptions
  ) => {
    await this._server.delete(path, handler, options)
  }

  notFound = async (handler: (requestDto: RequestDto) => Promise<Response>) => {
    await this._server.notFound(handler)
  }

  findAvailablePort = async (): Promise<number> => {
    return this._server.findAvailablePort()
  }

  start = async (port: number): Promise<number> => {
    return this._server.start(port)
  }

  stop = async () => {
    await this._server.stop()
  }
}
