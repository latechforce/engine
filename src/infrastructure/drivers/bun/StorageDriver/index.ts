import { PostgreSQLDriver } from '../../common/StorageDriver/PostgreSQLDriver'
import { SQLiteStorageDriver } from './SQLiteDriver'
import type { StorageConfig } from '/domain/services/Storage'
import type { IStorageDriver } from '/adapter/spi/drivers/StorageSpi'

export class StorageDriver implements IStorageDriver {
  private _storage: PostgreSQLDriver | SQLiteStorageDriver

  constructor({ driver, query, exec }: StorageConfig) {
    switch (driver) {
      case 'PostgreSQL':
        this._storage = new PostgreSQLDriver(query, exec)
        break
      case 'SQLite':
        this._storage = new SQLiteStorageDriver(query, exec)
        break
      default:
        throw new Error('Invalid driver')
    }
  }

  connect = async () => {
    await this._storage.connect()
  }

  disconnect = async () => {
    await this._storage.disconnect()
  }

  bucket = (name: string) => {
    return this._storage.bucket(name)
  }
}
