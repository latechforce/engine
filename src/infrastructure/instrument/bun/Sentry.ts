import type { MonitorSentryConfig } from '/domain/services/Monitor'
import * as Sentry from '@sentry/bun'
import slugify from 'slugify'

export function init(config: MonitorSentryConfig) {
  const release = `${slugify(config.appName, { lower: true })}@${config.appVersion}`
  Sentry.init({
    release,
    dsn: config.dsn,
    environment: config.environment,
    tracesSampleRate: 1.0,
  })
  console.log('Sentry has been successfully initialized')
}
