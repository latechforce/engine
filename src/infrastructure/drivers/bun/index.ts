import type { Drivers } from '/adapter/spi/drivers'
import type { DatabaseConfig } from '/domain/services/Database'
import type { MonitorsConfig } from '/domain/services/Monitor'
import type { ServerConfig } from '/domain/services/Server'
import type { StorageConfig } from '/domain/services/Storage'

import { DatabaseDriver } from './DatabaseDriver'
import { MonitorDriver } from './MonitorDriver'
import { ServerDriver } from './ServerDriver'

import { drivers as common } from '../common'
import { StorageDriver } from './StorageDriver'

export const drivers: Drivers = {
  ...common,
  monitor: (config: MonitorsConfig) => new MonitorDriver(config),
  database: (config: DatabaseConfig) => new DatabaseDriver(config),
  server: (config: ServerConfig) => new ServerDriver(config),
  storage: (config: StorageConfig) => new StorageDriver(config),
}
