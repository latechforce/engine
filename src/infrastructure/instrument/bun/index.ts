import type { ConfigSchema } from '/adapter/api/schemas/ConfigSchema'
import * as Sentry from './Sentry'
import { SystemDriver } from '/infrastructure/drivers/common/SystemDriver'

export function instrument(config: ConfigSchema) {
  const system = new SystemDriver()
  if (config.services?.monitors) {
    for (const monitor of config.services.monitors) {
      if (monitor.driver === 'Sentry')
        Sentry.init({
          appName: config.name,
          appVersion: config.appVersion ?? system.getAppVersion(),
          ...monitor,
        })
    }
  }
}
