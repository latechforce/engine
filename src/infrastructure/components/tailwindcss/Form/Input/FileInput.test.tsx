import { it, expect } from 'bun:test'
import { FileInput } from './FileInput'
import { renderToString } from 'react-dom/server'

it('should match the file input snapshot', async () => {
  // GIVEN
  const input = <FileInput field="name" multiple={false} />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the multiple file input snapshot', async () => {
  // GIVEN
  const input = <FileInput field="name" multiple={true} />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the file input snapshot with label', async () => {
  // GIVEN
  const input = <FileInput field="name" label="Name" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the file input snapshot with description', async () => {
  // GIVEN
  const input = <FileInput field="name" description="Description" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the file input snapshot with label and description', async () => {
  // GIVEN
  const input = <FileInput field="name" label="Name" description="Description" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the file input snapshot with placeholder', async () => {
  // GIVEN
  const input = <FileInput field="name" placeholder="Placeholder" />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})

it('should match the file input snapshot with required', async () => {
  // GIVEN
  const input = <FileInput field="name" required />

  // WHEN
  const html = renderToString(input)

  // THEN
  expect(html).toMatchSnapshot()
})
