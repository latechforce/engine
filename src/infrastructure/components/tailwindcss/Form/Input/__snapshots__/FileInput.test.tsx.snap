// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the file input snapshot 1`] = `"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name</label><label class="block"><span class="sr-only"></span><input type="file" class="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400" name="name"/></label></div>"`;

exports[`should match the multiple file input snapshot 1`] = `"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name</label><label class="block"><span class="sr-only"></span><input type="file" multiple="" class="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400" name="name"/></label></div>"`;

exports[`should match the file input snapshot with label 1`] = `"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">Name</label><label class="block"><span class="sr-only"></span><input type="file" class="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400" name="name"/></label></div>"`;

exports[`should match the file input snapshot with description 1`] = `"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name</label><p class="mb-2 text-sm text-gray-500 dark:text-neutral-500">Description</p><label class="block"><span class="sr-only"></span><input type="file" class="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400" name="name"/></label></div>"`;

exports[`should match the file input snapshot with label and description 1`] = `"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">Name</label><p class="mb-2 text-sm text-gray-500 dark:text-neutral-500">Description</p><label class="block"><span class="sr-only"></span><input type="file" class="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400" name="name"/></label></div>"`;

exports[`should match the file input snapshot with placeholder 1`] = `"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name</label><label class="block"><span class="sr-only">Placeholder</span><input type="file" class="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400" name="name"/></label></div>"`;

exports[`should match the file input snapshot with required 1`] = `"<div><label for="name" class="mb-2 block text-sm font-medium text-gray-700 dark:text-white">name<span class="text-red-500"> *</span></label><label class="block"><span class="sr-only"></span><input type="file" required="" class="block w-full text-sm text-gray-500 file:me-4 file:rounded-lg file:border-0 file:bg-blue-600 file:px-4 file:py-2 file:text-sm file:font-semibold file:text-white hover:file:bg-blue-700 file:disabled:pointer-events-none file:disabled:opacity-50 dark:text-neutral-500 dark:file:bg-blue-500 dark:hover:file:bg-blue-400" name="name"/></label></div>"`;
